src/tmpfiles/tmpfiles.c

```diff
@@ -2006,62 +2006,176 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
         return r;
 }
 
-static int rm_if_wrong_type(mode_t mode, const char *path, bool follow) {
+static int rm_if_wrong_type_safe(
+                mode_t mode,
+                int parent_fd,
+                const struct stat *parent_st /* Only used if follow is true. */,
+                const char *name,
+                bool follow) {
         struct stat st;
-        int r;
+        int r, fd;
+        const char *e;
 
-        assert(path);
+        assert(name);
         assert((mode & ~S_IFMT) == 0);
-        log_debug("Checking type of %s", path);
-
         if (follow)
-                r = stat(path, &st);
-        else
-                r = lstat(path, &st);
+                assert(parent_st);
+
+        e = strrchr(name, '/');
+        if (e)
+                return -EINVAL;
+
+        r = fstatat(parent_fd, name, &st, follow ? 0 : AT_SYMLINK_NOFOLLOW);
         if (r < 0) {
                 if (errno != ENOENT) {
-                        log_error_errno(r, "stat(%s): %m", path);
+                        // TODO fix this log to give full path like log_unsafe_transition
```

Comment:
I plan to do a pass to replace these TODOs with proper logging once the basic structure settles.